library/bignum.c

```diff
@@ -2395,6 +2395,31 @@ int mbedtls_mpi_gcd( mbedtls_mpi *G, const mbedtls_mpi *A, const mbedtls_mpi *B
     return( ret );
 }
 
+/* Fill X with n_bytes random bytes.
+ * X must already have room for those bytes.
+ * The size and sign of X are unchanged.
+ * n_bytes must not be 0.
+ */
+static int mpi_fill_random_internal(
+    mbedtls_mpi *X, size_t n_bytes,
+    int (*f_rng)(void *, unsigned char *, size_t), void *p_rng )
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    const size_t limbs = CHARS_TO_LIMBS( n_bytes );
+    const size_t overhead = ( limbs * ciL ) - n_bytes;
+
+    if( X->n < limbs )
+        return( MBEDTLS_ERR_MPI_BAD_INPUT_DATA );
+
+    memset( X->p, 0, overhead );
+    memset( (unsigned char *) X->p + limbs * ciL, 0, ( X->n - limbs ) * ciL );
+    MBEDTLS_MPI_CHK( f_rng( p_rng, (unsigned char *) X->p + overhead, n_bytes ) );
+    mpi_bigendian_to_host( X->p, limbs );
```

Comment:
This is not addressed it seems.

Reply:
Sorry, I forgot to reply. I don't understand your comment. We want to end up with random bytes up to the bound and zero bytes outside the bound, where “up to” is defined according to the value of the bignum. Since we're working with bytes, we need to rearrange the bytes accordingly. Note also that we want the bytes from the RNG in a specific order so that this function can be used as the PRNG for deterministic ECDSA. The endianness shuffling was preexisting, I believe that it was correct and I haven't knowingly changed how it works.

Reply:
I added a comment to state that the order in which the bytes are laid out here is deliberate.

Reply:
Looking closer at bignum.c and its big-endian/little-endian/host story it seems that `host` in `mpi_bigendian_to_host` rather means the little-endian representation of big numbers used in bignum.c rather the endianness of the host machine as `mpi_bigendian_to_host` does not check the machine endianness to reorder limbs. But then at limb level mpi_uint_bigendian_to_host does check the machine endianness ???

Reply:
After some discussions offline, the comment above is correct but for the "???" at the end. Yes, the limbs are uint32_t or uint64_t integers in host machine endianness.
And in my first comment:  "But if the host is big-endian this function shouldn't do anything". This is false, mpi_bigendian_to_host does not re-order the bytes constituting the limbs but re-orders the limbs in a little-endian way.